discretisedfield/tests/test_mesh.py

```diff
@@ -1173,43 +1173,155 @@ def test_coordinate_field(valid_mesh):  # TODO
+    [
+        [0, 100, list("a"), 10],
+        [(0, 0), (100, 50), list("ab"), (10, 10)],
+        [(0, 0, 0), (100, 60, 50), list("abc"), (10, 10, 10)],
+        [(0, 0, 0, 0), (100, 60, 50, 40), None, (10, 10, 10, 10)],
+    ],
+)
+def test_sel_range(p1, p2, dims, cell):
+    region = df.Region(p1=p1, p2=p2, dims=dims)
+    # Sel range of mesh with no subregions
+    mesh = df.Mesh(region=region, cell=cell)
+    for dim in mesh.region.dims:
+        options = {dim: (12.5, 29.5)}
+        sub_mesh = mesh.sel(**options)
+        assert isinstance(sub_mesh, df.Mesh)
+        assert sub_mesh.region.ndim == mesh.region.ndim
+        bool_ = [i != dim for i in mesh.region.dims]
+        idx = mesh.region._dim2index(dim)
+        assert np.isclose(sub_mesh.region.pmax[idx], 30.0, atol=0)
+        assert np.isclose(sub_mesh.region.pmin[idx], 10.0, atol=0)
+        assert np.allclose(sub_mesh.region.pmin[bool_], mesh.region.pmin[bool_], atol=0)
+        assert np.allclose(sub_mesh.region.pmax[bool_], mesh.region.pmax[bool_], atol=0)
+        assert sub_mesh.region.dims == mesh.region.dims
+        assert all(sub_mesh.n[bool_] == mesh.n[bool_])
+        assert sub_mesh.n[idx] == 2
+
+        # Single layer
+        options = {dim: (12.5, 13.5)}
+        sub_mesh = mesh.sel(**options)
+        assert isinstance(sub_mesh, df.Mesh)
+        assert sub_mesh.region.ndim == mesh.region.ndim
+        bool_ = [i != dim for i in mesh.region.dims]
+        idx = mesh.region._dim2index(dim)
+        assert np.isclose(sub_mesh.region.pmax[idx], 20.0, atol=0)
+        assert np.isclose(sub_mesh.region.pmin[idx], 10.0, atol=0)
+        assert np.allclose(sub_mesh.region.pmin[bool_], mesh.region.pmin[bool_], atol=0)
+        assert np.allclose(sub_mesh.region.pmax[bool_], mesh.region.pmax[bool_], atol=0)
+        assert sub_mesh.region.dims == mesh.region.dims
+        assert all(sub_mesh.n[bool_] == mesh.n[bool_])
+        assert sub_mesh.n[idx] == 1
+
+        # Too many values
+        with pytest.raises(ValueError):
+            options = {dim: (12.5, 29.5, 3.5)}
+            mesh.sel(**options)
+
+        # Out of bounds
+        with pytest.raises(ValueError):
+            options = {dim: (12.5, 1000)}
+            mesh.sel(**options)
+
+        # Wrong value type
+        with pytest.raises(TypeError):
+            options = {dim: (12.5, f"{dim}")}
+            mesh.sel(**options)
+
+
+def test_sel_subregions():
+    p1 = (0, 0, 0)
+    p2 = (20, 20, 20)
+    region = df.Region(p1=p1, p2=p2, dims=["x0", "x1", "x2"])
+    cell = (2, 2, 2)
+    mesh = df.Mesh(region=region, cell=cell)
+
+    sub_region = {
+        "in": df.Region(p1=(2, 2, 2), p2=(8, 8, 8)),
+        "out": df.Region(p1=(0, 0, 0), p2=(2, 2, 2)),
+        "half": df.Region(p1=(4, 4, 4), p2=(12, 12, 12)),
+    }
+    mesh = df.Mesh(region=region, cell=cell, subregions=sub_region)
+    sub_mesh = mesh.sel(x0=(3.6, 7.8))
+    assert [i for i in sorted(sub_mesh.subregions)] == ["half", "in"]
```

Comment:
I think this should work:
```suggestion
    assert sorted(sub_mesh.subregions) == ["half", "in"]
```